src/workerd/io/worker-fs.h

```diff
@@ -0,0 +1,192 @@
+#pragma once
+
+#include <workerd/jsg/url.h>
+
+#include <kj/filesystem.h>
+#include <kj/map.h>
+
+namespace workerd {
+
+// The FsMap is a configurable mapping of built-in "known" file system
+// paths to user-configurable locations. It is used to allow user-specified
+// "mount" points for certain built-in file system paths. For example, the
+// WorkerFileSystem exposes a built-in bundle path that is located at /bundle
+// by default. The FsMap allows the user to specify a different name for this
+// path, such as /mybundle, or even nested paths like /mybundle/a/b/c/modules/.
+class FsMap final {
+ public:
+  FsMap() = default;
+  KJ_DISALLOW_COPY_AND_MOVE(FsMap);
+
+  // The known roots are the special built-in paths that the FsMap can
+  // map to user-specified locations. To add a new known root, add the
+  // entry to this enum, specify the default path for the new root in
+  // the FsMap::Builder constructor, then configure the path by adding
+  // the necessary code in the buildDirectory(...) function in the
+  // WorkerFileSystem class defined in worker-fs.c++.
+  enum class KnownRoot {
+    // The root for the built-in bundle path. Every file in this path will
+    // be sourced from the worker configuration bundle. If not specified,
+    // the default root path is `/bundle`.
+    BUNDLE,
+    // The root for the built-in temp path.
+    TEMP,
+  };
+
+  // Gets the kj::Path mapped for the given known root.
+  kj::Path getPathForKnownRoot(KnownRoot root) const;
+
+  // Gets the base URL mapped for the given known root.
+  const jsg::Url& getUrlForKnownRoot(KnownRoot root) const;
+
+  // Returns the known root for the given path, if anything.
+  kj::Maybe<KnownRoot> tryGetKnownRootForUrl(const jsg::Url& path) const;
+
+  // Returns the known root for the given path, if anything.
+  kj::Maybe<KnownRoot> tryGetKnownRootForPath(kj::PathPtr path) const;
+
+  // Used to configure the FsMap instance, which is immutable once built.
+  class Builder final {
+   public:
+    Builder();
+
+    // The path must be parseable as a URL relative to the base URL "file:///".
+    // Only the path part of the URL is used.
+    void setPathForKnownRoot(KnownRoot root, kj::StringPtr path);
+    kj::Own<FsMap> finish();
+
+   private:
+    kj::Own<FsMap> map;
+  };
+
+  static kj::Own<FsMap> createDefault();
+
+ private:
+  kj::HashMap<KnownRoot, jsg::Url> paths;
```

Comment:
If `KnownRoot` is likely to stay finite and small (and is already an int-based value starting with 0) you'll be better off with a simple `kj::Array<kj::Maybe<jsg::Url>>(KnownRoot::MAX_VAL)`